p2p/tools/local_network.py

```diff
@@ -0,0 +1,378 @@
+
+    #
+    # Service API
+    #
+    async def _run(self) -> None:
+        while not self.cancel_token.triggered:
+            await asyncio.sleep(0.02)
+
+    async def _cleanup(self) -> None:
+        # all of the cancel tokens *should* be triggered already so we just
+        # wait for the networking processes to complete.
+        if self.connections:
+            await asyncio.wait(
+                self.connections.values(),
+                timeout=2,
+                return_when=asyncio.ALL_COMPLETED
+            )
+
+    #
+    # Connections API
+    #
+    def get_host(self, host: str) -> 'Host':
+        if host not in self.hosts:
+            self.hosts[host] = Host(host, self)
+        return self.hosts[host]
+
+    def get_network(self, name: str) -> 'Network':
+        if name not in self.networks:
+            self.networks[name] = Network(name, self)
+        return self.networks[name]
+
+    def get_connected_readers(self, address: Address) -> ReaderWriterPair:
+        external_reader, internal_writer = direct_pipe()
+        internal_reader, external_writer = addressed_pipe(address)
+
+        token = CancelToken(str(address)).chain(self.cancel_token)
+        connection = asyncio.ensure_future(_connect_streams(
+            internal_reader, internal_writer, external_writer.transport.queue, token,
+        ))
+        self.connections[token] = connection
+
+        return (external_reader, external_writer)
+
+
+ConnectionCallback = Callable[[asyncio.StreamReader, asyncio.StreamWriter], Any]
+
+
+class Server(asyncio.AbstractServer):
+    """
+    Mock version of `asyncio.Server` object.
+    """
+    connections: List[asyncio.StreamWriter]
+
+    def __init__(
+            self,
+            client_connected_cb: ConnectionCallback,
+            address: Address) -> None:
+        self.client_connected_cb = client_connected_cb
+        self.address = address
+        self.connections = []
+
+    def __repr__(self) -> str:
+        return '<%s %s>' % (self.__class__.__name__, self.address)
+
+    def close(self) -> None:
+        for writer in self.connections:
+            writer.write_eof()
+
+    async def wait_closed(self) -> None:
+        await asyncio.wait(
+            tuple(writer.drain() for writer in self.connections),
+            timeout=0.01,
+            return_when=asyncio.ALL_COMPLETED
+        )
+
+    def add_connection(self, writer: asyncio.StreamWriter) -> None:
+        self.connections.append(writer)
+
+
+class Network:
+    router: Router
+    name: str
+    _default_host: str = None
+
+    def __init__(self, name: str, router: Router, default_host: str=None):
+        self.name = name
+        self.router = router
+        self.default_host = default_host
+
+    @property
+    def default_host(self) -> str:
+        if self._default_host is not None:
+            return self._default_host
+        else:
+            return os.environ.get('P2P_DEFAULT_HOST', '127.0.0.1')
```

Comment:
This env variable name probably makes less sense when mock-network is in its own distro.